### DISCUSSION TITLE
Join queries aren't using same "root" instance

### DISCUSSION THREAD

**Discussion by Goldmensch** [micronaut-projects/micronaut-data#1258] [Q&A] [answered]
```java
@MappedEntity
public record Cart(
    @Id @GeneratedValue @Nullable
    Long id,
    String brand,
    @Relation(value = Kind.ONE_TO_MANY, mappedBy = "cart", cascade = Cascade.ALL)
    List<CartPart> parts) {}
```
```java
@MappedEntity()
public record CartPart(
    @Id @GeneratedValue @Nullable
    Long id,
    String type,
    @Relation(Kind.MANY_TO_ONE) @Nullable
    Cart cart
) {}
```
```java
@JdbcRepository
public interface TestRepo extends CrudRepository<Cart, Long> {
  @Override
  @Join(value = "parts")
  Optional<Cart> findById(@Id @Nullable Long id);
}
```

So when I fetch a Cart via `findById` then each CartPart has a different new instance of Cart and this instance hasn't any parts.
Is there any way to have Micronaut reuse the same instance in each CartPart?

test:
```java
    var cart = new Cart(null, "VW", List.of(
        new CartPart(null, "door", null),
        new CartPart(null, "motor", null)));

    var saved = repo.save(cart);
    var foundCart = repo.findById(saved.id()).orElseThrow();

    System.out.println(foundCart);
    System.out.println(foundCart.parts().get(0).cart());
    System.out.println(System.identityHashCode(foundCart));
    System.out.println(System.identityHashCode(foundCart.parts().get(0)));
    System.out.println(System.identityHashCode(foundCart.parts().get(1)));
```

output:
```
Cart[id=1, brand=VW, parts=[CartPart[id=1, type=door, cart=Cart[id=1, brand=VW, parts=[]]], CartPart[id=2, type=motor, cart=Cart[id=1, brand=VW, parts=[]]]]]
Cart[id=1, brand=VW, parts=[]]
1899145692
1464031233
765880900
Cart[id=1, brand=VW, parts=[CartPart[id=1, type=door, cart=Cart[id=1, brand=VW, parts=[]]], CartPart[id=2, type=motor, cart=Cart[id=1, brand=VW, parts=[]]]]]
```

**Comment by dstepanov** [comment, marked as answer]
You should probably post the output of:
```
System.out.println(System.identityHashCode(foundCart.parts().get(0).cart()));
System.out.println(System.identityHashCode(foundCart.parts().get(1).cart()));
```

When using immutable records you cannot simply modify the parent and have the children reference the updated parent.

**Reply by Goldmensch** [reply]
I'm sorry for being late, the output of this is:
```
1047000562
265629779
265629779
```
```java
    System.out.println(System.identityHashCode(foundCart));
    System.out.println(System.identityHashCode(foundCart.parts().get(0).cart()));
    System.out.println(System.identityHashCode(foundCart.parts().get(1).cart()));
```
So it seems that Micronaut only fetches/creates one new instance without any parts. That's a good message. But is there any way to let Micronaut use the parent instance already given, so that Micronaut uses the "foundCart" instance in the childs as well, and thus `foundCart` is the same as `foundCart.parts().get(n).cart()`?

**Reply by dstepanov** [reply]
> But is there any way to let Micronaut use the parent instance already given

It isn't possible when dealing with immutable structures because every modification requires a new instance. To set the parent into the child you need to modify the child and then you need to update the child instance in the parent etc

**Reply by Goldmensch** [reply]
> > But is there any way to let Micronaut use the parent instance already given
> 
> 
> 
> It isn't possible when dealing with immutable structures because every modification requires a new instance. To set the parent into the child you need to modify the child and then you need to update the child instance in the parent etc
> 
> 

well, seems logical. Thank you anyway
